zeus/monitor/carbon.py

```diff
@@ -0,0 +1,448 @@
+            )
+            resp = requests.get(url)
+        except requests.exceptions.RequestException as e:
+            logger.exception(
+                "Failed to retrieve current carbon intensity measurement: %s", e
+            )
+            raise
+
+        try:
+            return resp.json()["carbonIntensity"]
+        except KeyError as e:
+            # Raise exception when carbonIntensity does not exist in response
+            raise ZeusCarbonIntensityNotFoundError(
+                f"Current carbon intensity measurement not found at `({self.lat}, {self.long})` "
+                f"with estimate set to `{self.estimate}` and emission_factor_type set to `{self.emission_factor_type}`\n"
+                f"JSON Response: {resp.text}"
+            ) from e
+
+    def get_recent_carbon_intensity(self) -> dict[str, float]:
+        """Fetches recent (within last 24 hours) carbon intensity of the location of the class.
+
+        !!! Note
+                        In some locations, there is no recent carbon intensity data. `self.estimate` can be used to approximate the carbon intensity in such cases.
+        """
+        try:
+            url = (
+                f"https://api.electricitymap.org/v3/carbon-intensity/history?lat={self.lat}&lon={self.long}"
+                + f"&disableEstimations={not self.estimate}&emissionFactorType={self.emission_factor_type}"
+            )
+            resp = requests.get(url)
+        except requests.exceptions.RequestException as e:
+            logger.exception(
+                "Failed to retrieve recent carbon intensity measurement: %s", e
+            )
+            raise
+
+        try:
+            recent_carbon_intensities: dict[str, float] = {
+                measurement["datetime"]: measurement["carbonIntensity"]
+                for measurement in resp.json()["history"]
+            }
+            return recent_carbon_intensities
+        except KeyError as e:
+            # Raise exception when carbonIntensity does not exist in response
+            raise ZeusCarbonIntensityNotFoundError(
+                f"Recent carbon intensity measurement not found at `({self.lat}, {self.long})` "
+                f"with estimate set to `{self.estimate}` and emission_factor_type set to `{self.emission_factor_type}`\n"
+                f"JSON Response: {resp.text}"
+            ) from e
+
+
+@dataclass
+class CarbonEmissionMeasurement:
+    """Measurement result of one window.
+
+    Attributes:
+                    time: Time elapsed (in seconds) during the measurement window.
+                    gpu_energy: Maps GPU indices to the energy consumed (in Joules) during the
+                                    measurement window. GPU indices are from the DL framework's perspective
+                                    after applying `CUDA_VISIBLE_DEVICES`.
+                    gpu_carbon_emission: Maps GPU indices to the carbon emission produced (in mgCO2eq) during the
```

Comment:
Believe the units are correct, writing explanation here to verify:

ZeusMonitor.end_window returns Joules

when carbon intensity is calculated, I add a `/ 3600` to convert joules to Wh
`
                gpu_carbon_emissions[key][gpu_index] += (
                    energy_measurements[key][hour_key][f"gpu_{gpu_index}"]
                    / 3600
                    * carbon_intensity
                )
`

I then multiply Wh with the carbon intensity (gCO2eq/kWh == mgCO2eq / Wh since g = 1000 mg and kWh = 1000 Wh)

As for mgCO2eq, I wasn't really sure exactly the typical window length of users. When I ran very short windows (< 1 min), gCO2eq gave decimals values at the magnitude of around ~0.001

I was worried that because it was a floating point, there is floating imprecision that might make it lose accuracy? I made it mgCO2eq so that less of the significant digits are in the decimals. When running `run_clm.py`, 50 epochs of training on GPT2 (12 layers, 12 attention heads, 768 embedding size) took 20 minutes and yielded around 300,00 0 mgCO2eq per each of the 8 GPUs.

Do you happen to know how long users usually run these monitors for? Seems like if > 1-3 mins then we should change to gCO2eq? Happy to change it, just want to know your thoughts and give you the final say.

Reply:
Thanks for checking! I think `gCO2eq` is better. People will typically run the monitor for minutes, if not hours.